subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/execution/Combinators.kt

```diff
@@ -18,21 +18,69 @@ package org.gradle.kotlin.dsl.execution
 
 import org.jetbrains.kotlin.lexer.KotlinLexer
 import org.jetbrains.kotlin.lexer.KtToken
+import org.jetbrains.kotlin.lexer.KtTokens
 import org.jetbrains.kotlin.lexer.KtTokens.CLOSING_QUOTE
 import org.jetbrains.kotlin.lexer.KtTokens.COMMENTS
 import org.jetbrains.kotlin.lexer.KtTokens.IDENTIFIER
+import org.jetbrains.kotlin.lexer.KtTokens.LBRACE
+import org.jetbrains.kotlin.lexer.KtTokens.LBRACKET
 import org.jetbrains.kotlin.lexer.KtTokens.LPAR
 import org.jetbrains.kotlin.lexer.KtTokens.OPEN_QUOTE
+import org.jetbrains.kotlin.lexer.KtTokens.RBRACE
+import org.jetbrains.kotlin.lexer.KtTokens.RBRACKET
 import org.jetbrains.kotlin.lexer.KtTokens.REGULAR_STRING_PART
 import org.jetbrains.kotlin.lexer.KtTokens.RPAR
 import org.jetbrains.kotlin.lexer.KtTokens.SEMICOLON
 import org.jetbrains.kotlin.lexer.KtTokens.WHITE_SPACE
+import kotlin.reflect.KProperty
 
 
 internal
 typealias Parser<T> = KotlinLexer.() -> ParserResult<T>
 
 
+@Suppress("unused")
+private
+val debugger: ParserDebugger = ParserDebugger()
```

Comment:
🤔  I think I just saw global shared mutable state. Any chance we can avoid this?

Reply:
Have improved on it; please take a look (will soon link the follow-up PR).